script/tool/lib/src/publish_check_command.dart

```diff
@@ -121,24 +188,67 @@ class PublishCheckCommand extends PluginCommand {
             'Packages with an SDK constraint on a pre-release of the Dart SDK should themselves be published as a pre-release version.');
   }
 
-  Future<bool> _passesPublishCheck(Directory package) async {
+  Future<_PublishCheckResult> _passesPublishCheck(Directory package) async {
     final String packageName = package.basename;
     print('Checking that $packageName can be published.');
 
     final Pubspec pubspec = _tryParsePubspec(package);
     if (pubspec == null) {
-      return false;
+      print('no pubspec');
+      return _PublishCheckResult._error;
     } else if (pubspec.publishTo == 'none') {
       print('Package $packageName is marked as unpublishable. Skipping.');
-      return true;
+      return _PublishCheckResult._noPublish;
+    }
+
+    final Version version = pubspec.version;
+    final bool alreadyPublished = await _checkIfAlreadyPublished(
+        packageName: packageName, version: version);
+    if (alreadyPublished) {
+      print(
+          'Package $packageName version: $version has already be published on pub.');
+      return _PublishCheckResult._noPublish;
     }
 
     if (await _hasValidPublishCheckRun(package)) {
       print('Package $packageName is able to be published.');
-      return true;
+      return _PublishCheckResult._needsPublish;
     } else {
       print('Unable to publish $packageName');
-      return false;
+      return _PublishCheckResult._error;
     }
   }
+
+  // Check if `packageName` already has `version` published on pub.
+  Future<bool> _checkIfAlreadyPublished(
+      {String packageName, Version version}) async {
+    final PubVersionFinderResponse pubVersionFinderResponse =
+        await _pubVersionFinder.getPackageVersion(package: packageName);
+    bool published;
+    switch (pubVersionFinderResponse.result) {
+      case PubVersionFinderResult.success:
+        published = pubVersionFinderResponse.versions.contains(version);
+        break;
+      case PubVersionFinderResult.fail:
+        print(_machineMessageError);
+        printErrorAndExit(errorMessage: '''
+Error fetching version on pub for $packageName.
+HTTP Status ${pubVersionFinderResponse.httpResponse.statusCode}
+HTTP response: ${pubVersionFinderResponse.httpResponse.body}
+''');
```

Comment:
I'm realizing that failures here are going to be almost impossible to debug, because we won't have any information in the log.

Perhaps we should consider doing something like a JSON dictionary of output, with status and detailed log messages as separate keys? That would mean we'd need to consume the result with something like a Dart script that can parse the JSON, rather than just a one-liner in the GitHub Action. Thoughts?

Reply:
Good Idea, updated the code to do exactly that